File: packages/sdk-py/src/arckit/clients/usdc.py
"""USDC (ERC-20) client."""

from __future__ import annotations

from web3 import Web3

from arckit._tx import TxContext
from arckit.abi import ERC20_ABI
from arckit.utils import from_usdc_base, to_usdc_base


class UsdcClient:
    def __init__(self, ctx: TxContext, usdc_address: str):
        self._ctx = ctx
        self._contract = ctx.w3.eth.contract(
            address=Web3.to_checksum_address(usdc_address), abi=ERC20_ABI
        )

    def balance_of_raw(self, account: str) -> int:
        return self._contract.functions.balanceOf(Web3.to_checksum_address(account)).call()

    def balance_of(self, account: str) -> float:
        return from_usdc_base(self.balance_of_raw(account))

    def allowance(self, owner: str, spender: str) -> int:
        return self._contract.functions.allowance(
            Web3.to_checksum_address(owner), Web3.to_checksum_address(spender)
        ).call()

    def approve(self, spender: str, amount: float | int) -> dict:
        return self._ctx.send(
            self._contract.functions.approve(
                Web3.to_checksum_address(spender), to_usdc_base(amount)
            )
        )

    def transfer(self, to: str, amount: float | int) -> dict:
        return self._ctx.send(
            self._contract.functions.transfer(Web3.to_checksum_address(to), to_usdc_base(amount))
        )

File: packages/sdk/src/clients/usdc.ts
import type { Account, Address, PublicClient, WalletClient } from 'viem';
import { erc20Abi } from '../abi/erc20.js';
import { WalletRequiredError } from '../errors.js';
import { fromUsdcBase, toUsdcBase, waitForReceipt } from '../utils.js';

export type UsdcClientOpts = {
  publicClient: PublicClient;
  walletClient?: WalletClient;
  account?: Account;
  usdc: Address;
};

export class UsdcClient {
  constructor(private readonly opts: UsdcClientOpts) {}

  /// Get raw balance (base units, 6 decimals).
  async balanceOfRaw(account: Address): Promise<bigint> {
    return this.opts.publicClient.readContract({
      address: this.opts.usdc,
      abi: erc20Abi,
      functionName: 'balanceOf',
      args: [account],
    });
  }

  /// Get balance as a USDC float (e.g. 12.5 for 12.5 USDC).
  async balanceOf(account: Address): Promise<number> {
    return fromUsdcBase(await this.balanceOfRaw(account));
  }

  async allowance(owner: Address, spender: Address): Promise<bigint> {
    return this.opts.publicClient.readContract({
      address: this.opts.usdc,
      abi: erc20Abi,
      functionName: 'allowance',
      args: [owner, spender],
    });
  }

  async approve(spender: Address, amount: number | bigint) {
    const account = this.requireAccount('approve');
    const hash = await this.opts.walletClient!.writeContract({
      account,
      chain: null,
      address: this.opts.usdc,
      abi: erc20Abi,
      functionName: 'approve',
      args: [spender, toUsdcBase(amount)],
    });
    return waitForReceipt(this.opts.publicClient, hash);
  }

  async transfer(to: Address, amount: number | bigint) {
    const account = this.requireAccount('transfer');
    const hash = await this.opts.walletClient!.writeContract({
      account,
      chain: null,
      address: this.opts.usdc,
      abi: erc20Abi,
      functionName: 'transfer',
      args: [to, toUsdcBase(amount)],
    });
    return waitForReceipt(this.opts.publicClient, hash);
  }

  private requireAccount(op: string): Account {
    if (!this.opts.account || !this.opts.walletClient) {
      throw new WalletRequiredError(op);
    }
    return this.opts.account;
  }
}
